xdsl/dialects/experimental/halo.py

```diff
@@ -0,0 +1,341 @@
+            builtin.DenseArrayBase.from_list(typ, source_offset),
+            builtin.IntegerAttr.from_params(neighbor, typ),
+        ])
+
+    @property
+    def offset(self) -> tuple[int]:
+        data = self.offset_.as_tuple()
+        assert isa(data, tuple[int])
+        return data
+
+    @property
+    def size(self) -> tuple[int]:
+        data = self.size_.as_tuple()
+        assert isa(data, tuple[int])
+        return data
+
+    @property
+    def source_offset(self) -> tuple[int]:
+        data = self.source_offset_.as_tuple()
+        assert isa(data, tuple[int])
+        return data
+
+    @property
+    def neighbor(self) -> int:
+        data = self.neighbor_.value.data
+        assert isa(data, int)
+        return data
+
+    @property
+    def elem_count(self) -> int:
+        return prod(self.size)
+
+    @property
+    def dim(self) -> int:
+        return len(self.size)
+
+    def source_area(self) -> 'HaloExchangeDecl':
+        """
+        Since a HaloExchangeDef by default specifies the area to receive into,
+        this method returns the area that should be read from.
+        """
+        # we set source_offset to all zeor, so that repeated calls to source_area never return the dest area
+        return HaloExchangeDecl.from_tuples(
+            offset=tuple(
+                val + offs
+                for val, offs in zip(self.offset, self.source_offset)),
+            size=self.size,
+            source_offset=tuple(0 for _ in range(len(self.source_offset))),
+            neighbor=self.neighbor,
+        )
+
+    def print_parameters(self, printer: Printer) -> None:
+        printer.print_string('<at [')
+        printer.print_list(self.offset, lambda x: printer.print_string(str(x)))
+        printer.print_string('] size [')
+        printer.print_list(self.size, lambda x: printer.print_string(str(x)))
+        printer.print_string('] source offset [')
+        printer.print_list(self.source_offset,
+                           lambda x: printer.print_string(str(x)))
+        printer.print_string('] to {}>'.format(self.neighbor))
+
+
+@irdl_attr_definition
+class HaloShapeInformation(ParametrizedAttribute):
+    """
+    This represents shape information that is attached to halo operations.
+
+    On the terminology used:
+
+    In each dimension, we are given four points. We abbreviate them in
+    annotations to an, bn, cn, dn, with n being the dimension. In 2d, these
+    create the following pattern, higher dimensional examples can
+    be derived from this:
+
+    a0 b0          c0 d0
+    +--+-----------+--+ a1
+    |  |           |  |
+    +--+-----------+--+ b1
+    |  |           |  |
+    |  |           |  |
+    |  |           |  |
+    |  |           |  |
+    +--+-----------+--+ c1
+    |  |           |  |
+    +--+-----------+--+ d1
+
+    We can now name these points:
+
+         - a: buffer_start
+         - b: core_start
+         - c: core_end
+         - d: buffer_end
```

Comment:
Are `a` and `d` representing the bounds of the memory buffer, or the halo?
If it supposed to represent the halo, I find the name confusion-prone and would like to suggest renaming it to halo_start/halo_end or similar.